sql-plugin/src/main/scala/com/nvidia/spark/rapids/SplitColumnarBatchProcessor.scala

```diff
@@ -0,0 +1,333 @@
+      currentBatch.append((rowsInPartition, valuesInPartition))
+      partSchema.zipWithIndex.foreach {
+        case (field, colIndex) if field.dataType == StringType =>
+          val singleValue = valuesInPartition.getString(colIndex)
+          val sizeOfSingleValue = singleValue.getBytes("UTF-8").length.toLong
+          val sizeOfPartition = rowsInPartition * sizeOfSingleValue
+          // Check if adding the partition to the current batch exceeds cuDFLimit
+          if (sizeOfBatch.getOrElseUpdate(colIndex, 0L) + sizeOfPartition <= cuDF_LIMIT) {
+            sizeOfBatch(colIndex) += sizeOfPartition
+          } else {
+            // Need to split the current partition
+            val newRows = (cuDF_LIMIT - sizeOfBatch(colIndex)) / sizeOfSingleValue
+            // Update batches and clear states
+            currentBatch.update(currentBatch.length - 1, (newRows, valuesInPartition))
+            resultBatches.append(currentBatch.toArray)
+            currentBatch.clear()
+            sizeOfBatch.clear()
+            // Reprocess the current partition with remaining rows
+            partRows(partIndex) -= newRows
+            partIndex -= 1
+          }
+        case _ =>
+        // Assumption: Columns of other data type would not have fit already. Do nothing.
+      }
+      partIndex += 1
+    }
+    // Add the last remaining batch to resultBatches
+    if (currentBatch.nonEmpty) {
+      resultBatches.append(currentBatch.toArray)
+    }
+    resultBatches.toArray
+  }
+
+  /**
+   * Splits the input ColumnarBatch into smaller batches, wraps these batches with partition
+   * data, and returns an Iterator.
+   *
+   * Note:
+   * - Partition values are not converted to GPU ColumnVectors and merged with the smaller
+   *   ColumnarBatch to avoid wasting GPU memory.
+   * - The actual creation of GPU ColumnVectors and merging is handled by the
+   *   SplitBatchIteratorWithPartitionData lazily.
+   *
+   * @param batch           Input ColumnarBatch.
+   * @param splitRowData    Array of arrays containing split row data (RowNums and InternalRows).
+   * @param partitionSchema Schema of partition ColumnVectors.
+   * @return An Iterator of combined ColumnarBatches.
+   */
+  private def splitAndCombineBatchWithPartitionData(
+      batch: ColumnarBatch,
+      splitRowData: Array[Array[(Long, InternalRow)]],
+      partitionSchema: StructType): Iterator[ColumnarBatch] = {
+    withResource(batch) { _ =>
+      val splitIndices = calculateSplitIndices(splitRowData)
+      val splitColumnarBatches = splitColumnarBatch(batch, splitIndices)
+      // Combine the split GPU ColumnVectors with partition ColumnVectors.
+      val combinedBatches = splitColumnarBatches.zip(splitRowData).map {
+        case (splitColumnarBatch, splitRowInfo) =>
+          (0 until splitColumnarBatch.numCols).map(splitColumnarBatch.column).foreach {
+            case gpuCol: GpuColumnVector => gpuCol.incRefCount()
```

Comment:
Why are we incrementing ref counts here?  If new batches were created because of splitting, this causes leaks.